dojo/engagement/views.py

```diff
@@ -85,36 +85,62 @@ def engagement_calendar(request):
 
 
 def engagement(request):
-    products = get_authorized_products(Permissions.Engagement_View).distinct()
-    if request.user.is_staff:
-        engagements = Engagement.objects.all()
-    else:
-        engagements = Engagement.objects\
-            .filter(Q(product__authorized_users=request.user) | Q(product__prod_type__authorized_users=request.user))\
-            .distinct()
+    engagements = Engagement.objects.filter(
+        product__in=get_authorized_products(Permissions.Engagement_View).distinct()
+    ).select_related(
+        'product',
+        'product__prod_type',
+    ).prefetch_related(
+        'lead',
+        'tags',
+        'product__tags',
+    )
 
-    products_with_engagements = products.filter(~Q(engagement=None), engagement__active=True).distinct()
-    filtered = EngagementFilter(
+    # Get the test counts per engagments. As a separate query, this is much
+    # faster than annotating the above `engagements` query.
+    engagement_test_counts = {
+        test['engagement']: test['test_count']
+        for test in Test.objects.filter(
+            engagement__in=engagements
+        ).values(
+            'engagement'
+        ).annotate(
+            test_count=Count('engagement')
+        )
+    }
+
+    if System_Settings.objects.get().enable_jira:
+        engagements = engagements.prefetch_related(
+            'jira_project__jira_instance',
+            'product__jira_project_set__jira_instance'
+        )
+
+    filtered_engagements = EngagementDirectFilter(
         request.GET,
-        queryset=products_with_engagements.prefetch_related('engagement_set', 'prod_type', 'engagement_set__lead',
-                                                            'engagement_set__test_set__lead', 'engagement_set__test_set__test_type'))
-    prods = get_page_items(request, filtered.qs, 25)
-    name_words = products_with_engagements.values_list('name', flat=True)
-    eng_words = engagements.filter(active=True).values_list('name', flat=True).distinct()
+        queryset=engagements
+    )
+    engs = get_page_items(request, filtered_engagements.qs, 25)
+    product_name_words = sorted(set(
+        engagement.product.name
+        for engagement in engs.object_list
+    ))
+    engagement_name_words = sorted(set(
+        engagement.name
+        for engagement in engs.object_list
+    ))
```

Comment:
not sure, but I think the name words are now limited to the engagements on the current page only?

Reply:
Oh yeah you're right, I got overzealous removing queries. Will fix.